crates/project/src/terminals.rs

```diff
@@ -18,63 +19,83 @@ impl Project {
     pub fn create_terminal(
         &mut self,
         working_directory: Option<PathBuf>,
+        spawn_runnable: Option<SpawnRunnable>,
         window: AnyWindowHandle,
         cx: &mut ModelContext<Self>,
     ) -> anyhow::Result<Model<Terminal>> {
-        if self.is_remote() {
-            return Err(anyhow::anyhow!(
-                "creating terminals as a guest is not supported yet"
-            ));
-        } else {
-            let settings = TerminalSettings::get_global(cx);
-            let python_settings = settings.detect_venv.clone();
-            let shell = settings.shell.clone();
-
-            let terminal = TerminalBuilder::new(
-                working_directory.clone(),
-                shell.clone(),
-                settings.env.clone(),
-                Some(settings.blinking.clone()),
-                settings.alternate_scroll,
-                window,
+        anyhow::ensure!(
+            !self.is_remote(),
+            "creating terminals as a guest is not supported yet"
+        );
+
+        let settings = TerminalSettings::get_global(cx);
+        let python_settings = settings.detect_venv.clone();
+        let (completion_tx, completion_rx) = bounded(1);
```

Comment:
Nitpick: we could probably use `futures::channel::oneshot` instead here.

Reply:
The problem with that one that it takes `self` to resolve and is not `Clone` — so we have to fiddle with `Option` and design things more complex.
But in this particular case this all might actually work, since it's a small, internal thing.

Reply:
Ah yeah, if there's a good reason to have it then let's keep it. `bounded(1)` just struck me as a bit odd.

Reply:
Had tried replacing those quickly and found that currently the problem will be in 
https://github.com/zed-industries/zed/blob/c731a08432b6db6ff9309a8f9a88917df08e7f3b/crates/terminal/src/terminal.rs#L1389-L1392

At any given moment, when some long-to-complete runnables are running, the user can schedule another one to spawn — and, with the certain settings, that runnable will have to wait for the original one to complete.
Right now the code is made so that all future spawns replace the previous wait task with the new one, and it's not really possible with `Option::take` approach that `oneshot` forces us to do — maybe there are ways to chain those tasks, but the code around will become much harder.

So, while I agree that the current version is somewhat bad, esp. from the styling point of view, but I'd keep it.